File: src/native_test_lib/ios/XpmTestLib/XpmTestLib/XpmTestLib.swift
//
//  XpmTestLib.swift
//  XpmTestLib
//
//  Created by Mike Parker on 18/12/2020.
//

import Foundation
import XamarinPlatformMessaging

@objc(XpmTestLib)
public class XpmTestLib : NSObject {
    
    static var _channel : MessageChannel?
    
    @objc(start)
    public static func start() {
        _channel = MessageChannel(channelId: "xpm_test_lib")
        
        if let channel = _channel {
            channel.setHandler(forMessageId: "log_hello_text") { (parameter) in
                NSLog("Hello from Swift")
                sendMessage(messageId: "log_hello_text_executed", value: "Hello from Swift")
            }
            
            channel.setHandler(forMessageId: "log_custom_text") { (parameter) in
                if let params = parameter, let arg = params.first as? NSString {
                    NSLog(arg.description)
                    sendMessage(messageId: "log_custom_text_executed", value: arg.description)
                }
                else {
                    NSLog("No text provided")
                    sendMessage(messageId: "log_custom_text_executed", value: "No text provided")
                }
            }
            
            channel.setResponseHandler(forMessageId: "get_text") { (parameter) in
                return "Hello from Swift" as NSObject
            }
            
            channel.setResponseHandler(forMessageId: "get_text_with_param") { (parameter) in
                if let params = parameter, let arg = params.first as? NSString {
                    return "Parameter received: \(arg)" as NSObject
                }
                
                return "No parameter received" as NSObject
            }
        }
        
        sendMessage(messageId: "start_executed")
    }
    
    @objc(stop)
    public static func stop() {
        guard let channel = _channel else {return}
        
        do {
            try channel.clearHandler(forMessageId: "log_hello_text")
            try channel.clearHandler(forMessageId: "log_custom_text")
            try channel.clearResponseHandler(forMessageId: "get_text")
            try channel.clearResponseHandler(forMessageId: "get_text_with_param")
        }
        catch {
            NSLog(error.localizedDescription)
        }
        
        sendMessage(messageId: "stop_executed")
        _channel = nil
    }
    
    static func sendMessage(messageId: String) {
        sendMessage(messageId: messageId, value: nil)
    }
    
    static func sendMessage(messageId: String, value: String?) {
        if let channel = _channel {
            do {
                if let arg = value {
                    try channel.send(messageId: messageId, parameters: arg as NSObject)
                }
                else {
                    try channel.send(messageId: messageId)
                }
            } catch {
                NSLog(error.localizedDescription)
            }
        }
    }
}

File: src/native_test_lib/android/XpmTestLib/xpm-test-lib/src/main/java/com/mobcat/xpmtestlib/XpmTestLib.java
package com.mobcat.xpmtestlib;

import android.content.Context;
import android.util.Log;

import com.xamarin.platformmessaging.MessageChannel;

public class XpmTestLib {
    static Context _context;
    static MessageChannel _channel;

    public static void start(Context context) {
        _channel = new MessageChannel("xpm_test_lib");
        _context = context;

        _channel.setHandler("log_hello_text", (parameter) -> {
            Log.d("XpmTestLib", "Hello from Java");
            sendMessage("log_hello_text_executed", "Hello from Java");
        });

        _channel.setHandler("log_custom_text", (parameter) -> {
            String arg = (parameter != null && parameter.length == 1) ? (String)parameter[0] : null;

            if (arg != null) {
                Log.d("XpmTestLib", arg);
                sendMessage("log_custom_text_executed", arg);
            }
            else {
                Log.d("XpmTestLib", "No text provided");
                sendMessage("log_custom_text_executed", "No text provided");
            }
        });

        _channel.setResponseHandler("get_text", (parameter) -> {
            return "Hello from Java";
        });

        _channel.setResponseHandler("get_text_with_param", (parameter) -> {
            String arg = (parameter != null && parameter.length == 1) ? (String)parameter[0] : null;

            if (arg != null) {
                return String.format("Parameter received: %s", arg);
            }

            return "No parameter received";
        });

        sendMessage("start_executed");
    }

    public static void stop() {
        try {
            _channel.clearHandler("log_hello_text");
            _channel.clearHandler("log_custom_text");
            _channel.clearResponseHandler("get_text");
            _channel.clearResponseHandler("get_text_with_param");

            sendMessage("stop_executed");
        }
        catch (Exception ex) {
            Log.d("XpmTestLib", ex.getLocalizedMessage());
        }

        _channel = null;
        _context = null;
    }

    static void sendMessage(String messageId) {
        sendMessage(messageId, null);
    }

    static void sendMessage(String messageId, String value) {
        try {
            if (value != null) {
                _channel.send(messageId, value);
            }
            else {
                _channel.send(messageId);
            }
        }
        catch (Exception ex) {
            Log.d("XpmTestLib", ex.getLocalizedMessage());
        }
    }
}